src/day8/re_frame/trace/components.cljs

```diff
@@ -13,3 +15,32 @@
     [:title "remove"]
     [:path
       {:d "M31.708 25.708c-0-0-0-0-0-0l-9.708-9.708 9.708-9.708c0-0 0-0 0-0 0.105-0.105 0.18-0.227 0.229-0.357 0.133-0.356 0.057-0.771-0.229-1.057l-4.586-4.586c-0.286-0.286-0.702-0.361-1.057-0.229-0.13 0.048-0.252 0.124-0.357 0.228 0 0-0 0-0 0l-9.708 9.708-9.708-9.708c-0-0-0-0-0-0-0.105-0.104-0.227-0.18-0.357-0.228-0.356-0.133-0.771-0.057-1.057 0.229l-4.586 4.586c-0.286 0.286-0.361 0.702-0.229 1.057 0.049 0.13 0.124 0.252 0.229 0.357 0 0 0 0 0 0l9.708 9.708-9.708 9.708c-0 0-0 0-0 0-0.104 0.105-0.18 0.227-0.229 0.357-0.133 0.355-0.057 0.771 0.229 1.057l4.586 4.586c0.286 0.286 0.702 0.361 1.057 0.229 0.13-0.049 0.252-0.124 0.357-0.229 0-0 0-0 0-0l9.708-9.708 9.708 9.708c0 0 0 0 0 0 0.105 0.105 0.227 0.18 0.357 0.229 0.356 0.133 0.771 0.057 1.057-0.229l4.586-4.586c0.286-0.286 0.362-0.702 0.229-1.057-0.049-0.13-0.124-0.252-0.229-0.357z"}]])
+
+(defn scroll! [el start end time]
+  (.play (fx/Scroll. el (clj->js start) (clj->js end) time)))
+
+(defn scrolled-to-end? [el tolerance]
+  ;; at-end?: element.scrollHeight - element.scrollTop === element.clientHeight
+  (> tolerance (- (.-scrollHeight el) (.-scrollTop el) (.-clientHeight el))))
+
+(defn autoscroll-list [{:keys [class scroll?] :as opts} childr]
+  (let [should-scroll (r/atom true)]
+    (r/create-class
+      {:display-name "autoscroll-list"
+       :component-did-mount
+                     (fn [this]
+                       (let [n (r/dom-node this)]
```

Comment:
I'm pretty sure that using a _callback ref_ is the preferred way to get a reference to the DOM element over `r/dom-node` nowadays:
https://github.com/reagent-project/reagent/issues/251
https://presumably.de/reagent-mysteries-part-3-manipulating-the-dom.html

Seems to be more future-proof and safe according to the React team :)

Reply:
I believe the pattern would be something like

```
  (let [should-scroll (r/atom true)
        n (atom nil)]
...
```

```
       :reagent-render
                     (fn [{:keys [class]} child]
                       [:div {:class class :ref (fn [dom-node] (reset! n dom-node))} child])})))
```

And then the dom-node can be accessed in the other lifecycly functions with `@n`.

Reply:
thanks, that's great feedback! I'll adjust the code